### ISSUE TITLE
Accessibility issues in the default template

### ISSUE THREAD

**Issue by saschaleib** [dokuwiki/dokuwiki#4349]
This is an attempt to clarify and extend on the various accessibility issues that are found e.g. in the [WCAG Audit](https://www.dokuwiki.org/wcag-audit) page (with a special hat-tip for the good work by @selfthinker), but also the automated reports by [Wave](https://wave.webaim.org/) and [Lighthouse](https://developer.chrome.com/docs/lighthouse/) on Chromium-based browsers.

Note that this is not an exhaustive list of accessibility issues. This focuses on those issues that are easy to test (especially, but not only, with the above mentioned tools) and can be fixed with small interventions. The aim is to establish compliance with [WCAG 2.2, level AA](https://www.w3.org/TR/WCAG22/) for the structural elements and selected content items (as found in the "Formatting Syntax" page. Other page structures, and additional tools (such as the media manager, etc.) are outside of scope for now and may be looked at later.

**Colour contrast:** Some commonly used combinations of default colours do not have sufficient contrast with pre-defined background colours (as a reminder, a minimum of 4.5:1 is required to match [WCAG 1.3.1 Contrast](https://www.w3.org/WAI/WCAG22/quickref/?versions=2.1&showtechniques=131#contrast-minimum).

The following is an overview of the text colours' contrast against the various background colours:

| Variable        | default   | `@ini_background`  | `@ini_background_alt` | `@ini_background_neu` |
| --------        | --------  | -----------------  | --------------------- | --------------------- |
|                 |           | (=`#FFFFFF`)       | (=`#EEEEEE`)          | (=`#DDDDDD`)          |
| `@ini_text`     | `#333333` | 12.63:1 (OK)       | 10.88:1 (OK)          | 9.30:1 (OK)           |
| `@ini_text_alt` | `#999999` |  **2.84:1** (fail) |  **2.45:1** (fail)    | **2.09:1** (fail)     |
| `@ini_text_neu` | `#666666` |  5.74:1 (OK)       |  4.94:1 (OK)          | **4.22:1** (fail)     |
| `@ini_link`     | `#2b73b7` |  4.95:1 (OK)       |  **4.27:1** (fail)    | **3.64:1** (fail)     |
| `@ini_existing` | `#008800` |  4.64:1 (OK)       |  **4.00:1** (fail)    | **3.41:1** (fail)     |
| `@ini_missing`  | `#dd3300` |  4.61:1 (OK)       |  **3.97:1** (fail)    | **3.39:1** (fail)    |

To summarise: many combinations of text and background colours have insufficient contrast. In most of these cases, a small change is all that is needed.

While **border colours** are generally not an issue for accessibility, the `@ini_border` (`#CCCCCC`) value is used also as a fill colour for icons (including for the user- and the page-tools). As it has been noted in #3443 as the first point, this leads to insufficient contrast (1.6:1).

The code areas use the `@ini_background_site` colour as background (default: `#fbfaf9`). Some of the syntax highlighting colours have insufficient contrast against this background. This applies to the following colours as defined in the `geshi.less` file:

| selector(s) | default   | contrast    | 
| ---         | ---       | ---         |
| `.nu0`      | `#cc66cc` |  **3.17:1** | 
| `.re4`      | `#009999` |  **3.34:1** |
| `.re8`      | `#00b000` |  **2.79:1** |
| `.br0`, `.sy0`               | `#66cc66` |  **1.93:1** |
| `.co1`, `.coMULTI`, `.sc-1 ` | `#808080` |  **3.78:1** |
| `.kw1`, `.kw8`               | `#b1b100` |  2.19:1     |
| `.re1`, `.st0`, `.st_h `     | `#ff0000` |  3.83:1     |

Certain input fields and some other form elements have hard-coded colours instead of making use of the LESS variables defined in `style.ini`. This also leads to insufficient contrast e.g. in the search field – potentially even worse in situations where the site admin decided to alter these colours. 

**Links must be distinguishable without relying on colour alone** (see [WCAG 1.4.1 Use of Color](https://www.w3.org/TR/WCAG22/#use-of-color). This rule is met by the _missing_ links (underlined), as well as for _external_ and _interwiki_ links (by adding an icon). However, this is not true for the regular _existing_ (green) links, which have no attribute to distinguish them from the regular text except for their colour.

**Size issues:** All elements should have sufficient sizes, so (a) the text remains readable, and (b) they can be operated in a touch-screen environment:

- A number of UI elements have a reduced font size in relation to the already quite small base font-size. Notably the breadcrumbs end up rather small and difficult to use. This should probably be considered in the context of a general revision of font-sizes (see also: "Other issues" below)

- The Search field is rather small and difficult to activate. This applies especially for the "search" button, which is a plain image button with a fixed size of `19px` × `14px`, probably because of the size of the image used here.

- the line spacing in the TOC, as well as in the search popup is very tight, making it difficult to activate a specific item, especially on touch devices.

**Providing semantic article structure:** As it is, each wiki page contains a single _article_ which is subdivided into one or more _sections_, each indicated with a headline (matching [WCAG Technique H69](https://www.w3.org/WAI/WCAG22/Techniques/html/H69)). The HTML code is however missing the opportunity to indicate these semantics by using `<article>` and `<section>` elements and uses non-semantic `<div>`s instead.

In addition, the TOC provides an in-page navigation list. As such, it should be marked up with a `<nav>` element.

**ARIA roles:** The TOC also uses [aria-expanded](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded) to indicate its list part is in the "hidden" state. However, this is not the purpose of this attribute. The correct aria attribute to use here is [aria-hidden](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden), which is explicitly intended for "Offscreen or collapsed content, such as menus".

**Logical headings level structure:** While the structure of the headlines in the content is outside of the control of the template, it should be made sure that the template sections themselves have a logical headlines structure (see also: [WCAG Technique G141](https://www.w3.org/WAI/WCAG22/Techniques/general/G141)). This affects especially the User tools headline (`<h3 class="a11y">User Tools</h3>`), which follows immediately after an `<h1>` headline and should therefore be a `<h2>` instead.

It would be worth discussing if the page logo should really be an `H1` headline, or of this level should not be reserved for the page content. However, in either case, the User Tools label should be `H2`.

**Headings and labels** should be added to all interactive elements (see [WCAG SC 2.4.5: Headings and Labels](https://www.w3.org/WAI/WCAG22/Understanding/headings-and-labels.html)). These are missing for the search field and the mobile version of the tools menu. In both cases, it is sufficient to add an `aria-label` attribute.

Other issues

- The template uses `accesskey`s for some functionality. These can interfere with keyboard shortcuts that add-on software (such as accessibility tools) bring to a user client. These should still be reviewed and at least be reduced to those where the benefit clearly outweights the potential for problems.

- The base font size is relatively small, which makes much of the text difficult to read. This also has the effect that the standard _measure_ (number of characters per line) in desktop view ends up as a very high 150en (recommended is about 80-85 max). It should be considered to use a larger base font size and adjust non-content elements accordingly.

- The template has no systematic approach to providing focus information. An easy approach here would be to use the outline CSS properties to make sure a focus border is drawn consistently (using a designated focus colour) around the currently focused link or form element.

**Comment by saschaleib** [comment]
As a PS: here is a complete list of all the syntax highlighting colours that are defined in the file `geshi.css`:

| selector(s) | default   | contrast    | Result | Notes|
| ---         | ---       | ---         | ---    | --- |
| `.co0`      | `#666666` |  5.50:1     | OK     | |
| `.co4`      | `#cc0000` |  5.64:1     | OK     | |
| `.es5`      | `#006699` |  5.99:1     | OK     | Bold |
| `.es6`      | `#009933` |  **3.59:1** | OK     | Bold |
| `.kw2`      | `#000000` | 20.14:1     | OK     | Bold |
| `.kw5`      | `#008000` |  4.92:1     | OK     | |
| `.kw6`      | `#FF0088` |  **3.59:1** | OK     | Bold |
| `.me0`      | `#004000` | 11.62:1     | OK     | |
| `.nu0`      | `#cc66cc` |  **3.17:1** | _fail_ | |
| `.re0`      | `#0000ff` |  8.24:1     | OK     | |
| `.re3`      | `#ff3333` |  5.64:1     | OK     | Bold |
| `.re4`      | `#009999` |  **3.34:1** | _fail_ | |
| `.re5`      | `#660033` | 12.49:1     | OK     | |
| `.re7`      | `#991111` |  8.24:1     | OK     | |
| `.re8`      | `#00b000` |  **2.79:1** | _fail_ | |
| `.sc-2`     | `#404040` |  9.94:1     | OK     | |
| `.sy3`      | `#000040` | 18.75:1     | OK     | |
| `.br0`, …   | `#66cc66` |  **1.93:1** | _fail_ | |
| `.co1`, …   | `#808080` |  **3.78:1** | _fail_ | |
| `.co2`, …   | `#2D842D` |  4.52:1     | OK     | |
| `.co3`, …   | `#008080` |  4.57:1     | OK     | |
| `.es0`, …   | `#000099` | 13.79:1     | OK     | Bold |
| `.es2`, …   | `#737373` |  9.94:1     | OK     | Bold |
| `.kw1`, …   | `#b1b100` |  2.19:1     | _fail_ | |
| `.kw10`, …  | `#003399` | 10.41:1     | OK     | Bold |
| `.kw13`, …  | `#006600` |  6.94:1     | OK     | |
| `.kw3`, …   | `#000066` | 16.9:1      | OK     | |
| `.kw4`, …   | `#993333` |  7.00:1     | OK     | |
| `.re1`, …   | `#ff0000` |  3.83:1     | _fail_ | |

Note that Bold elements are not consider a fail, if they still have at least a 3:1 contrast ratio.
